src/ripple/beast/clock/basic_seconds_clock.cpp

```diff
@@ -0,0 +1,103 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of Beast: https://github.com/vinniefalco/Beast
+    Copyright 2021, Howard Hinnant <howard.hinnant@gmail.com>
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/beast/clock/basic_seconds_clock.h>
+
+#include <chrono>
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+
+namespace beast {
+
+namespace {
+
+// Updates the clock
+class seconds_clock_thread
+{
+    using Clock = basic_seconds_clock::Clock;
+
+    bool stop_;
+    std::mutex mut_;
+    std::condition_variable cv_;
+    std::thread thread_;
+    Clock::time_point tp_;
+
+public:
+    ~seconds_clock_thread();
+    seconds_clock_thread();
+
+    Clock::time_point
+    now();
+
+private:
+    void
+    run();
+};
+
+seconds_clock_thread::~seconds_clock_thread()
+{
+    if (thread_.joinable())
+    {
+        {
+            std::lock_guard lock(mut_);
+            stop_ = true;
+        }
+        cv_.notify_one();
```

Comment:
Based on previous conversations with @JoelKatz, this should go inside the locked section.

Reply:
Agreed with you (and David) in general.  But in this case there's a very minor advantage to releasing the lock as soon as possible.  The waiting thread is doing a timed-wait, and has the opportunity to time-out every second.  If the time-out happens after the stop is published, but before the notify happens, then we see a small optimization in the waiting thread.  The predicate is satisfied, the notification becomes superfluous, and the shutdown happens just slightly faster.

All that being said, if you (or anyone) feels strongly enough about putting the notify (but not the join) inside the lock, I'll move it.  It won't make a huge difference either way.

Reply:
If we leave the code as is it might be nice to include a comment describing the benefit of the current structure.